firedrake/slate/slac/kernel_builder.py

```diff
@@ -380,3 +391,429 @@ def needs_mesh_layers(self):
+
+        self.needs_cell_orientations = False
+        self.needs_cell_sizes = False
+        self.needs_cell_facets = False
+        self.needs_mesh_layers = False
+
+        # TSFC calls only for terminal tensors
+        terminal_tensors = [tensor for tensor in var2terminal.values() if isinstance(tensor, slate.Tensor)]
+        tsfc_kernels = self.context_kernels(terminal_tensors)
+
+        # For all terminal tensors provided by TSFC
+        for pos, cxt_kernel in enumerate(tsfc_kernels):
+            integral_type = cxt_kernel.original_integral_type
+            tensor = cxt_kernel.tensor
+            loopy_tensor = tensor2temp[tensor]
+            mesh = tensor.ufl_domain()
+
+            if integral_type not in self.supported_integral_types:
+                raise ValueError("Integral type '%s' not recognized" % integral_type)
+
+            for tsfc_kernel in cxt_kernel.tsfc_kernels:
+                kinfo = tsfc_kernel.kinfo
+                reads = []
+                inames_dep = []
+
+                # Populate subkernel call to tsfc
+                templated_subkernels.append(kinfo.kernel.code)
+
+                # Prepare lhs and args for call to tsfc kernel
+                output = self.generate_lhs(tensor, pym.Variable(loopy_tensor.name))
+                kernel_data = self.collect_tsfc_kernel_data(mesh, cxt_kernel.coefficients, kinfo)
+                reads.extend(self.loopify_tsfc_kernel_data(kernel_data))
+
+                # Generate predicates for different integral types
+                if self.is_integral_type(integral_type, "cell_integral"):
+                    predicates = None
+                    if kinfo.subdomain_id != "otherwise":
+                        raise NotImplementedError("No subdomain markers for cells yet")
+                elif self.is_integral_type(integral_type, "facet_integral"):
+                    predicates, fidx, facet_arg = self.facet_integral_predicates(mesh, integral_type, kinfo)
+                    reads.append(facet_arg)
+                    inames_dep.append(fidx[0].name)
+                elif self.is_integral_type(integral_type, "layer_integral"):
+                    predicates = self.layer_integral_predicates(tensor, integral_type)
+                else:
+                    raise ValueError("Unhandled integral type {}".format(integral_type))
+
+                # TSFC kernel call
+                key = integral_type+"_tsfc_kernel_call%d" % len(assembly_calls[integral_type])
+                call = pym.Call(pym.Variable(kinfo.kernel.name), tuple(reads))
+                insn = loopy.CallInstruction((output,), call,
+                                             within_inames=frozenset(inames_dep),
+                                             predicates=predicates, id=key)
+                assembly_calls[integral_type].append(insn)
+
+        self.assembly_calls = assembly_calls
+        self.templated_subkernels = templated_subkernels
```

Comment:
I can't yet follow what's going on here. This class is very stateful and I think is likely still quite delicate. As I understand it, for each terminal object in `var2terminal` (where the terminal is a slate Tensor that wraps either a UFL form or a coefficient) you need to generate the sequence of loopy instructions that initialise that variable. You also need to collect (somehow) the extra loopy kernels that will be required.

Some state must be maintained across the terminal filling, because they might need the same sets of coefficients. However, most of it can be done independently, so I think I would recommend trying to do something like:


1. generate mapping from split coefficients of the outer slate expression to loopy GlobalArgs.
2. write a function that takes this mapping along with a single entry in var2terminal and returns the instructions and kernels necessary to initialise this terminal

Then we can just:

```python
instructions, kernels = zip(*(generate_initialisation(var, terminal, coeff_mapping) for var, terminal in var2terminal.items())
```

I think this is then more maintainable, because inside `generate_initialisation` you can dispatch straight away on the type of the terminal (assembledvector/cell integral/etc...) and have single-use functions for each of those.

I guess the other thing we need to determine is what "extra" information the outer wrapping kernel will need (e.g. cell2facets etc...). But I think that can be handled by inspecting either the initial terminals, or else returning `enum` values from `generate_initialisation` as well.

WDYT?

I guess I'm glossing over some bits an pieces, since we also need to pass extra info in (like the layer_arg and so forth). However, I think the best way to do this is just unilaterally cons this information in a small object "bag" and provide it as some parameters to `generate_initialisation`.

Reply:
With kernels you mean the the loopy calls to the TSFC kernels? Or do you mean the actual TSFC kernels?

Reply:
The tsfc kernels. Since i think the instructions contain the loopy calls

Reply:
I might still not quite understand what you mean, but here my thoughts.

The main work of the kernel builder is in `generate_tsfc_calls`. Retrieving the TSFC kernel is done within a few lines within `context_kernels()`, to loopify them takes more effort (loopyfing=generating a call to those kernels). The work that we need to do is:
- generate the data structure to take in the result of the tsfc call (`generate_lhs`)
- inspect which data the TSFC kernel reads in `collect_tsfc_kernel_data`. Since the data is in array form, we need to make the data loopy-readable with `SubarrayRef ` in `loopify_tsfc_kernel_data`
- inspect the type of integrals that the TSFC calls contain, so we can generate the right predicates (`is_integral_type`,  `facet_integral_predicates`, `layer_integral_predicates`).

We could pull `generate_tsfc_calls` out of the `KernelBuilder` and control all of the above from `merge_loopy` to expose more of the kernel-call-generation-process. But I think, from my view of the functionality, it should stay in the `KernelBuilder`, because it is literally **builds** the calls to the TSFC **kernels**. I don't want to pollute `merge_loopy` with that whole loopify dance. I want to keep it clear in there, that the slate wrapper kernel contains instructions
- to zero the tensor
- fill the tensors via TSFC calls
- do algebra operations on them,

but I don't really want to expose there how we generate those instructions.

Potentially, we can do a similar thing to your suggestion inside `generate_tsfc_calls` in the `KernelBuilder` in order to handle the different integral types nicer. I don't quite know yet - I am still thinking about this.

Independent of that, I think, pulling out the collection of the coefficients is possible and a good idea, I only do it once upfront and it does not depend on the tsfc calls.

And yes, with your suggestion we would need to spill the information about `layer_arg`s and co back as well.

Reply:
I absolutely agree with your wish to keep the `merge_loopy` frontend logic basically as it looks now. I think that I'm concerned about how we, under the hood, do the second step of your three-step process (filling the tensors via TSFC calls).

I think what I'm trying to understand is to what extent the second two steps of your work to collect data and so forth (inspect data, inspect integrals) interdepend on loopyfying other tsfc kernels. I think they basically don't once you've gathered/packed all the coefficients.

Part of this is hard to see because everything lives in the same class and so we have to read all the functions to know what bits they access and so forth.

I agree that a certain amount of messiness is difficult to avoid, because the loopy representation is kind of difficult to generate from a level in which indices are not explicit (this was why in the pyop2 codegen I introduced an IR between "parloop" and loopy). I think we don't have to go that far here.

What I also see, reading the code, is that there is a lot of `if isinstance(...)` dispatch based on the types/shapes of internal datastructures. This smells fragile to me: I can understand why we might have to dispatch on the basis of "external" data, but why must we do so for datastructures we control?

For example, I think I would prefer that the index creation/management were something like:

```python
from collections import OrderedDict
from itertools import count

from pymbolic import primitives as pym
from tsfc.loopy import create_domains


class IndexCreator(object):
    def __init__(self, prefix="i"):
        self.index_map = OrderedDict()  # pym variable -> extent
        self.namer = (f"{prefix}{c}" for c in count())

    def indices(self, *extents):
        """Create new indices with specified extents.

        :arg extents. Extent of each index
        :returns: tuple of pymbolic Variable objects representing
            indices, one for each extent."""
        indices = []
        for name, extent in zip(self.namer, extents):
            self.index_map[name] = extent
            indices.append(pym.Variable(name))
        return tuple(indices)

    @property
    def domains(self):
        """ISL domains for the currently known indices."""
        return create_domains(self.index_map.items())
```

We still need something to hold on to this thing, but now it's clearer that index creation is only in one small place.

Reply:
I don't think that there is a dependence **between** the code generation of the tsfc kernels. They all need the coefficients of the slate expression, yes. Additionally, as soon as one of the tsfc calls needs a predicate, the variables used in that predicate (like the number of layers) must be fed as an argument to the slate wrapper kernel. That has no effect on the calls to the other tsfc kernels, however.

Reply:
Okay two more thoughts. 
1) To make it more clear what data is accessed, I could put the variables which are needed for the construction of a call to a TSFC kernels in a bag, so e.g. which variables the kernels reads, which predicates it needs & co. Then we could map from each terminal to its own bag.
2) Then, in `generate_tsfc_kernels` rather than explicitly looping over the kernels we could do something similar to what you suggested
```
terminal2tsfckernel = zip(terminals, get_tsfc_kernels(terminal))
terminal2output = zip(terminals, construct_lhs(terminal))
terminal2data = zip(terminals, get_tsfc_kernel_data(terminal))
terminal2args = zip(terminals, loopy_kernel_data(terminal)) 
terminal2predicates = ...
```
Each of the functions (second arguments in zip above) would need to eat the terminal and would need access to the corresponding bag, so we need to keep a `terminal2bag` mapping as class property or pass the whole bag as an arg to those functions.

In fact, we might not even need the `terminal2tsfckernel` as mappings (only the terminal2bag) as long as they are in an ordered data structure and the data in it is grouped by terminal (not 100 percent sure about that yet).

I think one disadvantage of this might be, that we potentially have many bag objects dependent on how many tensors the slate algebra operations act on. You would also still need to actually have a look in that function to see which variables of the bag it alters, so maybe this is still not quite what we want.

Would this be a a bit closer to your suggestion?

Reply:
I think this feels closer. Can we flip it round so that we have a single function that takes a terminal and (because we think that the terminals don't need to share mutable state) a single _immutable_ bag of shared things (like the symbols for the layer arg, etc...) and returns the loopy instructions and tsfc-generated loopy kernel(s) for filling that terminal?

So something like

```python
bag = ...
bag.indices = IndexCreator()
bag.various_symbols = ...
bag.coefficient2symbol = ...

def generate(terminal, bag):
    do_whatever
    return instructions, kernels

instructions, kernels = zip(*generate(terminal, bag) for terminal in terminals)
domains = bag.indices.domains
big_kernel = loopy.make_kernel(domains, instructions + linalg_instructions)
for k in kernels:
    big_kernel.register_...(k)
```

I guess we need to provide some instruction dependencies between instructions and linalg_instructions

What am I missing here? I guess the arguments to the big kernel?

Reply:
Okay so this is different to what I suggested. What I was suggesting was concerning the code in the loop here https://github.com/firedrakeproject/firedrake/blob/fe23f0ef298e7a3683c064e53ed9131eee2e7b54/firedrake/slate/slac/kernel_builder.py#L781 Your bag is containing information needed for the `slate_wrapper_kernel`/`big_kernel`, mine for the construction of the tsfc loopy call instructions.

But, I have finally understood what you mean with instructions and kernels. You are unhappy about the fact that I am appending everything to the lists `templated_subkernels` and `assembly_calls` , owned by the kernel builder here https://github.com/firedrakeproject/firedrake/blob/fe23f0ef298e7a3683c064e53ed9131eee2e7b54/firedrake/slate/slac/kernel_builder.py#L816 You want the items of them to be returned separately. I think I can do that, the only thing to consider here is that compilation of one terminal with tsfc was delivering multiple kernel calls at some point. I think that was, because we were splitting the form - but not anymore, so the innermost loop in `generate_tsfc_calls` might be superfluous anyways. 

So I think I can give that a go now.

- I would make the bag not immutable however, then I can also add the information if we need extra arguments like mesh layers (which is information I get from the tsfc kernels). I need the information to generate the argument to the "big kernel" here https://github.com/firedrakeproject/firedrake/blob/fe23f0ef298e7a3683c064e53ed9131eee2e7b54/firedrake/slate/slac/utils.py#L333, so this function would need access to the bag as well. 
- Apart from that the `merge_loopy` could look similar to your sketch above, only "generate" should definitely be part of the `KernelBuilder`. The bag can be generated in the `merge_loopy` method as in your sketch (alternatively it could be a class property of the `KernelBuilder`.

Apart from that, would it still make sense to have another bag, one for the data needed for the generation of the instruction for tsfc kernel of each terminal?

In regards to what you have missed above. You basically sketched `merge_loopy` again, comparing the sketch to `merge_loopy` reveals what is missing. The call here https://github.com/firedrakeproject/firedrake/blob/fe23f0ef298e7a3683c064e53ed9131eee2e7b54/firedrake/slate/slac/utils.py#L345 is what you would want to do for the big_kernel.

Reply:
> I think I can do that, the only thing to consider here is that compilation of one terminal with tsfc was delivering multiple kernel calls at some point.

You also get multiple kernel calls if you have (say) `Tensor(foo*dx + bar*ds)`

Reply:
Ah yes, but I actually don't think it matters, it just means I need to zip correctly in the end. What do you think about the rest of my comment?

Reply:
I think this looks reasonable. Let's try and flesh things out in code in a little more detail without trying to get everything right at once. Can we get this structure in place and see if we're missing anything?

Reply:
Yes, I think I can come up with something tomorrow.

Reply:
Okay here a draft

- I have encapsulated the data in a slate wrapper bag now, so the data looks a bit more sorted. The object is owned by the KernelBuilder. I am not convinced I want to change that, because a) I don't really want to pass the bag around between function calls and b) I also don't want to return it from merge_loopy to the compiler, so that I can the generate the KernelInfo from the bag here https://github.com/firedrakeproject/firedrake/blob/2a01c30c3d1331b24b1fa7d49ae203ff2450f0eb/firedrake/slate/slac/compiler.py#L162
... though it still adressess your concern about the coefficients. That the coefficients are collected independently from the generation of the tsfc kernel calls becomes clearer through this line https://github.com/firedrakeproject/firedrake/blob/2a01c30c3d1331b24b1fa7d49ae203ff2450f0eb/firedrake/slate/slac/utils.py#L331 and where they are used here https://github.com/firedrakeproject/firedrake/blob/2a01c30c3d1331b24b1fa7d49ae203ff2450f0eb/firedrake/slate/slac/utils.py#L337
- I return the instructions and kernel in generate. And I do that per context kernel (for different integral types) per terminal, zipping everything together here https://github.com/firedrakeproject/firedrake/blob/2a01c30c3d1331b24b1fa7d49ae203ff2450f0eb/firedrake/slate/slac/utils.py#L339
- I still want to include the `IndexCreator` I am not yet convinced that the `IndexCreator` should be owned by the bag, and not by the `KernelBuilder` itself. I am sure the inames do.